onion_transport.go

```diff
@@ -74,6 +76,62 @@ type OnionTransport struct {
 	keysDir     string
 	keys        map[string]*rsa.PrivateKey
 	onlyOnion   bool
+	upgrader    *tptu.Upgrader
+}
+
+// Dial initializes a new connection to a peer at a given address
+func (t *OnionTransport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tpt.Conn, error) {
+	dialer, err := t.controlConn.Dialer(t.auth)
+	if err != nil {
+		return nil, err
+	}
+	netaddr, err := manet.ToNetAddr(raddr)
+	var onionAddress string
+	if err != nil {
+		onionAddress, err = raddr.ValueForProtocol(ma.P_ONION)
+		if err != nil {
+			return nil, err
+		}
+	}
+	onionConn := OnionConn{
+		transport: tpt.Transport(t),
+		raddr:     &raddr,
+	}
+	if onionAddress != "" {
+		split := strings.Split(onionAddress, ":")
+		onionConn.Conn, err = dialer.Dial("tcp4", split[0]+".onion:"+split[1])
+	} else {
+		onionConn.Conn, err = dialer.Dial(netaddr.Network(), netaddr.String())
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return t.upgrader.UpgradeOutbound(ctx, t, &onionConn, p)
+}
+
+// CanDial determines whether or not the transport can dial a given address
+func (t *OnionTransport) CanDial(addr ma.Multiaddr) bool {
+	if t.onlyOnion {
+		// only dial out on onion addresses
+		return IsValidOnionMultiAddr(addr)
+	}
+	return IsValidOnionMultiAddr(addr) || mafmt.TCP.Matches(addr)
+}
+
+// Protocols returns only the Onion protocol if onlyOnion is true, otherwise
+// it returns Onion and TCP.
+func (t *OnionTransport) Protocols() []int {
+	protocols := []int{ma.P_ONION}
+	if !t.onlyOnion {
+		protocols = append(protocols, ma.P_TCP)
+	}
+	return protocols
+}
+
+// Proxy returns true, for tor is a relay-based protocol.
+func (t *OnionTransport) Proxy() bool {
+	return true
```

Comment:
This should actually be false. Proxy is used when the rest of the multiaddr should be handled by some remote server.

This is an unfortunate consequence of how we dial. Ideally, we'd dial from left-to-right across the multiaddr, building up a connection from the ground up. However, this turned out to be really painful to implement (in go).

Reply:
ahhhhh, i see.